App/CheckpointStore.fs

```diff
@@ -0,0 +1,58 @@
+module CheckpointStore
+
+open Equinox
+open Propulsion.Feed
+
+module Events =
+  type Event =
+    | Checkpoint of {| pos: int64 |}
+    interface TypeShape.UnionContract.IUnionContract
+
+  let codec = FsCodec.SystemTextJson.Codec.Create<Event>()
+
+module Fold =
+  type State = int64 option
+  let initial = None
+
+  let evolve _ =
+    function
+    | Events.Checkpoint x -> Some x.pos
+
+  let fold: State -> Events.Event seq -> State = Seq.fold evolve
+
+let streamId = StreamId.gen2 SourceId.toString id
+type CheckpointService internal (resolve: string -> StreamId -> Equinox.Decider<Events.Event, Fold.State>) =
```

Comment:
I wasn't in the mood to add another PG user and schema for the checkpoints. This just checkpoints into an adjacent stream.

e.g. if you're group `MyGroup` listening to the category `MyCategory`, it'll checkpoint to `MyCategory:position-messageDb_MyGroup`. The `messageDb` comes from propulsion and is to allow for multiple sources.

Reply:
Might upstream this to propulsion

Reply:
I'd not be surprised if @bartelink had already made a storage agnostic decider like this that I'm just not aware of

Reply:
The document-based stores (which can do rollingstate updates) do this https://github.com/jet/propulsion/blob/master/src/Propulsion.CosmosStore/ReaderCheckpoint.fs
But an equivalent for PG is not currently implemented (might not be the daftest idea!)
For now I think using https://github.com/jet/propulsion/blob/master/src/Propulsion.MessageDb/ReaderCheckpoint.fs#L49 is simply the right thing to do, even if it makes some mess ?

Reply:
wrt upstreaming, having an ever-growing stream would be a blocker

Also while in general it can be helpful to keep checkpoints in the same place as the info being derived (and that's the case here), another consideration is that you want to avoid the feedback effect of writing a checkpoint position triggering a (null) reaction (and checkpoint update every 5s) each time

For messagedb,  if you keep Invoice, InvoiceNumber and the checkpoints in the same DB, [as events] this would be the case
- if MessageDb positions are per category then this does not apply
- if you kept all 3 categories in ESDB it would (global pos would inc after you write a checkpoint pos)
- for cosmos, changefeed checkpoints are kept in a `-aux` separated container to avoid this effect
- for dynamo, the Indexer _should_ exclude 'system streams' such as `$ReaderCheckpoint-{x}` https://github.com/jet/propulsion/blob/1d862e907502d727f5508b401deb85c6c60ef1e7/src/Propulsion.CosmosStore/ReaderCheckpoint.fs#L17 at https://github.com/jet/propulsion/blob/master/src/Propulsion.DynamoStore.Indexer/Handler.fs#L24 (logging issue) to avoid the equivalent (not this is not as critical as it may seem as there is only one event per hour)

Reply:
> For now I think using `MessageDb.ReaderCheckpoint` is simply the right thing to do

It's definitely not a bad idea. But as always, there are trade-offs.

- Needs a separate connection string (checkpoints are stored in a different schema and the message_store role only has access to the message_store).
- Can lead to unpredictable performance because PG needs to VACUUM the table every now and then

I would definitely default to the checkpoints table, but I think there's value in having the option to "just use messagedb"

> another consideration is that you want to avoid the feedback effect of writing a checkpoint position triggering a (null) reaction (and checkpoint update every 5s) each time

I don't think this applies to MessageDB as we keep the position on a per category basis. There's no feedback because the checkpoints go to a separate category.

> having an ever-growing stream would be a blocker

The position of message-db's authors (which I agree with) is that message db should not be in the business of deleting data. Any production deployment of message-db will include pruning procedures outside of the application. With that in mind I don't think it should be a blocker as it's entirely consistent with real-world usage of MessageDB.

Reply:
>With that in mind I don't think it should be a blocker as it's entirely consistent with real-world usage of MessageDB.

Not suggesting there should be built-in pruning.

>I don't think this applies to MessageDB as we keep the position on a per category basis. There's no feedback because the checkpoints go to a separate category.

If the position is the event index within the category and not affected by writes to other categories, I guess that will dramatically reduce the amount of checkpointing (but e.g. if you wanted to store ESDB checkpoints in PG, then an event every 5s is going to add up, and inevitably someone will get paged at 5am and get annoyed with @nordfjord when they `git blame`!)

>I would definitely default to the checkpoints table, but I think there's value in having the option to "just use messagedb"

Maybe for the sample have the checkpoints DB connstr be optional iff people opt into "unbounded stream" checkpoints ?

IOW if they run with `--checkpointStream`, they can avoid rigging the checkpoints schema ?

Maybe I'm being a bit strong - In the context of the sample as a whole, the aggregate does serve as a nice example impl. But I do feel it's a step too far to put it in the box unless it's in a `Internal` namespace with a name that conveys that it will grow forever and you might just want to use the one that keeps one row in the DB ;)

Reply:
> If the position is the event index within the category and not affected by writes to other categories, I guess that will dramatically reduce the amount of checkpointing

The position is from the global sequence. The position sequence in the category is therefore monotonic but not atomic (e.g. 1,3,8,12222 is a perfectly valid sequence of positions). Also note that the subscription is polling `get_category_messages` for each of the categories supplied (so we'll only get evens for the subscribed categories). This is of course different from how the other stores function where the subscription is to an `$all` stream.

To be clear, when I say "real-world" I mean that this is how [eventide](https://github.com/eventide-project/consumer-postgres/blob/master/lib/consumer/postgres/position_store.rb#L14-L19) does its checkpointing